### PR TITLE
Optionally retain custom field types

### PR DESCRIPTION
**Pull request by christiand93**
- Add setting to retain custom field types when using cloud logging format
- Update documentation accordingly
- Fix test cases failing due to async callback called too early

### CODE REVIEW

File: docs/general-usage/04-custom-fields.md [hunk 1]
```diff
@@ -10,31 +10,51 @@ permalink: /general-usage/custom-fields
...
 // ... "msg":"Test data" 
-// ... "#cf": {"string": [{"k":"field","v":"value","i":"0"}]}...
+// ... "#cf": {"string": [{"k":"field","v":"42","i":"0"}]}...
 ```
 
-## General use
+### SAP Cloud Logging
 
-As of version 6.5.0 this library will automatically detect which logging service your app is bound to and will set the logging format accordingly.
+When using the library with SAP Cloud Logging custom fields get added as top-level fields.
```
Comment (nicklas-dohrn) [inline docs/general-usage/04-custom-fields.md:44]:
```suggestion
When using the library with SAP Cloud Logging, custom fields get added as top-level fields.
```

File: docs/general-usage/04-custom-fields.md [hunk 2]
```diff
@@ -10,31 +10,51 @@ permalink: /general-usage/custom-fields
...
+Registering custom fields is not required in this case.
 
-For local testing purposes you can still enforce a specific format like this:
+By default all custom field values get stringified, which can help to prevent field type mismatches when indexing to SAP Cloud Logging.
+However, if you want to retain the original type of your custom fields you can change the type conversion configuration accordingly:
```
Comment (nicklas-dohrn) [inline docs/general-usage/04-custom-fields.md:48]:
```suggestion
However, if you want to retain the original type of your custom fields, you can change the type conversion configuration accordingly:
```

File: src/lib/logger/recordFactory.ts [hunk 3]
```diff
@@ -110,19 +111,23 @@ export default class RecordFactory {
...
-
             if ([CustomFieldsFormat.CloudLogging, CustomFieldsFormat.All, CustomFieldsFormat.Default].includes(customFieldsFormat)
                 || record.payload[key] != null || this.config.isSettable(key)) {
+                // Stringify, if conversion type 'stringify' is selected and value is not a string already.
+                if (customFieldsTypeConversion == CustomFieldsTypeConversion.Stringify && (typeof value) != "string") {
+                    record.payload[key] = jsonStringifySafe(value);;
+                } else {
                     record.payload[key] = value;
-                   
+                }
```
Comment (nicklas-dohrn) [inline src/lib/logger/recordFactory.ts:121]:
```suggestion
                }
```

File: src/test/acceptance-test/custom-fields.test.js [hunk 4]
```diff
@@ -122,6 +122,37 @@ describe('Test custom fields', function () {
...
+
+        describe('retain field types', function () {
+            beforeEach(function () {
+                log = importFresh("../../../build/main/index");
+                log.setCustomFieldsTypeConversion("retain");
+                log.setCustomFields({ "field-a": 42, "field-b": true, "field-c": { "key": "value" }});
+                log.logMessage("info", "test-message");
+            });
+
+            it('logs custom fields as strings', function () {
```
Comment (nicklas-dohrn) [inline src/test/acceptance-test/custom-fields.test.js:148]:
```suggestion
            it('logs custom fields with their retained type', function () {
```

File: src/test/unit-test/config.test.js [hunk 5]
```diff
@@ -100,6 +100,20 @@ describe('Test Config class', function () {
         })
     });
 
+    describe('Test setCustomFieldsTypeConversion', function () {
+        beforeEach(function () {
+            singleton.setCustomFieldsTypeConversion("retain");
+            result = singleton.getConfig();
+        });
+        it('sets custom fields format', function () {
```
Comment (nicklas-dohrn) [inline src/test/unit-test/config.test.js:108]: if the test is called set custom fields format, than the changing of the type conversion should be done in the test and not in the before and after methods.

File: src/lib/logger/recordFactory.ts [hunk 6]
```diff
@@ -110,19 +111,23 @@ export default class RecordFactory {
...
-            // Stringify, if necessary.
-            if ((typeof value) != "string") {
-                value = jsonStringifySafe(value);
-            }
-
             if ([CustomFieldsFormat.CloudLogging, CustomFieldsFormat.All, CustomFieldsFormat.Default].includes(customFieldsFormat)
                 || record.payload[key] != null || this.config.isSettable(key)) {
+                // Stringify, if conversion type 'stringify' is selected and value is not a string already.
+                if (customFieldsTypeConversion == CustomFieldsTypeConversion.Stringify && (typeof value) != "string") {
+                    record.payload[key] = jsonStringifySafe(value);;
```
Comment (nicklas-dohrn) [inline src/lib/logger/recordFactory.ts:118]:
```suggestion
                    record.payload[key] = jsonStringifySafe(value);
```

File: src/lib/logger/recordFactory.ts [hunk 7]
```diff
@@ -110,19 +111,23 @@ export default class RecordFactory {
 
         let indexedCustomFields: any = {};
         providedFields.forEach((value, key) => {
-            // Stringify, if necessary.
-            if ((typeof value) != "string") {
-                value = jsonStringifySafe(value);
-            }
-
             if ([CustomFieldsFormat.CloudLogging, CustomFieldsFormat.All, CustomFieldsFormat.Default].includes(customFieldsFormat)
```
Comment (nicklas-dohrn) [inline src/lib/logger/recordFactory.ts:114]: this whole block looks a little bit overcomplicated, and duplicates the stringify solution.
I am not sure if there is room to get this simplified, but maybe factoring out the stringify to a second method would at least get rid of the duplication

**Review by nicklas-dohrn — CHANGES_REQUESTED** [review]
The actual function looks a little bit crowded.
Besides that, only some minor oversights.

File: src/test/unit-test/config.test.js (hunk 5 above)
Comment (christiand93) [reply src/test/unit-test/config.test.js:108]: Actually, the `it` is an essential part of the description:
"it sets custom fields format" is not referring to the test code but to the code under test.

File: src/lib/logger/recordFactory.ts (hunk 7 above)
Comment (christiand93) [reply src/lib/logger/recordFactory.ts:114]: The structure of the two sections is similar, but the logic is different. 
We could skip the type check though. I would expect `jsonStringifySafe()` to be a no-op if the input is already a string.

**Review by nicklas-dohrn — APPROVED** [review]
LGTM

**Pull request merged**